mobile/examples/speech_recognition/android/app/src/main/java/ai/onnxruntime/example/speechrecognition/MainActivity.kt

```diff
@@ -0,0 +1,169 @@
+        Log.e(TAG, "Error: ${exception.localizedMessage}", exception)
+        runOnUiThread {
+            statusText.text = "Error"
+            resultText.text = exception.localizedMessage
+        }
+    }
+
+    private fun hasRecordAudioPermission(): Boolean =
+        ActivityCompat.checkSelfPermission(
+            this,
+            Manifest.permission.RECORD_AUDIO
+        ) == PackageManager.PERMISSION_GRANTED
+
+    override fun onRequestPermissionsResult(
+        requestCode: Int,
+        permissions: Array<out String>,
+        grantResults: IntArray
+    ) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+        if (requestCode == RECORD_AUDIO_PERMISSION_REQUEST_CODE) {
+            if (!hasRecordAudioPermission()) {
+                Toast.makeText(
+                    this,
+                    "Permission to record audio was not granted.",
+                    Toast.LENGTH_SHORT
+                ).show()
+            }
+        }
+    }
+
+    private fun resetDefaultAudioButtonState() {
+        runOnUiThread {
+            usePrerecordedAudioButton.isEnabled = true
+            recordAudioButton.isEnabled = true
+            stopRecordingAudioButton.isEnabled = false
+        }
+    }
+
+    private fun disableAudioButtons() {
+        runOnUiThread {
+            usePrerecordedAudioButton.isEnabled = false
+            recordAudioButton.isEnabled = false
+            stopRecordingAudioButton.isEnabled = false
+        }
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+
+        usePrerecordedAudioButton.setOnClickListener {
+            // Disable audio buttons first.
+            // The audio button state will be reset at the end of the use prerecorded audio task.
+            disableAudioButtons()
+
+            workerThreadExecutor.submit {
+                try {
+                    val audioTensor = resources.openRawResource(R.raw.audio_mono_16khz_f32le).use {
+                        AudioTensorSource.fromRawPcmBytes(it.readBytes())
+                    }
+                    val result = audioTensor.use { speechRecognizer.run(audioTensor) }
+                    setSuccessfulResult(result)
+                } catch (e: Exception) {
+                    setError(e)
+                } finally {
+                    resetDefaultAudioButtonState()
+                }
+            }
+        }
+
+        recordAudioButton.setOnClickListener {
+            if (!hasRecordAudioPermission()) {
+                requestPermissions(
+                    arrayOf(Manifest.permission.RECORD_AUDIO),
+                    RECORD_AUDIO_PERMISSION_REQUEST_CODE
+                )
+                return@setOnClickListener
+            }
+
+            // Disable audio buttons first.
+            // The stop button will be enabled by the recording task.
+            disableAudioButtons()
```

Comment:
wondering why do we need this "disable" state handling of the audio button?

Reply:
since pressing this button starts some work in another thread, I didn't want to allow it to be pressed again before the recording is done and have to worry about keeping state consistent.